Optimize log2 tables

keszybz opened the pull request.

bluca commented on src/basic/util.h:
```diff
@@ -22,32 +22,35 @@ void in_initrd_force(bool value);
 int on_ac_power(void);
 
-static inline unsigned log2u64(uint64_t n) {
+#define LOG2U64(x) (x > 1 ? (unsigned) __builtin_clzll(x) ^ 63U : 0)
```
double eval of `x`?

keszybz replied:
It turns out that if we use a block ({ }) to avoid the double eval, this does not work as a compile-time constant anymore. I added a comment instead.

medhefgo replied:
You could write it like this:
```c
#define CONST_LOG2ULL(x) (x > 1 ? (unsigned) __builtin_clzll(x) ^ 63U : 0)
#define NONCONST_LOG2ULL(x) ({ int _x = x; _x > 1 ? (unsigned) __builtin_clzll(_x) ^ 63U : 0; })
#define LOG2ULL(x) __builtin_choose_expr(__builtin_constant_p(x), CONST_LOG2ULL(x), NONCONST_LOG2ULL(x))
```

keszybz replied:
Well, OK. But it's only executed on constants… I think this is too complex to implement "just in case". If the need ever arises, we can use it.

poettering replied:
i was about to propose the name, i.e. a single macro tat covers both cases. Less chance of using incorrectly...

bluca reviewed (approved):
lgtm, just one comment

yuwata commented on src/network/networkd-link.c:
```diff
@@ -2673,37 +2673,34 @@ DEFINE_STRING_TABLE_LOOKUP(link_state, LinkState);
+                [LOG2U(IFF_DYNAMIC)] =     "dynamic"       , /* dialup device with changing addresses. */
+                [LOG2U(IFF_LOWER_UP)] =    "lower-up"      , /* driver signals L1 up. */
+                [LOG2U(IFF_DORMANT)] =     "dormant"       , /* driver signals dormant. */
+                [LOG2U(IFF_ECHO)] =        "echo"          , /* echo sent packets. */
```
Suprious alignment of `=` and `,`...

yuwata commented on src/network/networkd-address.c:
```diff
@@ -26,33 +26,30 @@
-                                &str, ",",
-                                map[i].flag == IFA_F_SECONDARY && family == AF_INET6 ? "temporary" : map[i].name))
-                        return -ENOMEM;
+                if (FLAGS_SET(flags, 1 << i) && map[i])
```
`UINT32_C(1)`?

keszybz replied:
I don't think that's useful. By default we get `int`, and if we were to go outside of the range, the compiler will warn.

yuwata replied:
> if we were to go outside of the range, the compiler will warn.

Does it? If so, that's ok.

keszybz replied:
It does...
```
../src/network/networkd-address.c:48:40: error: left shift count >= width of type [-Werror=shift-count-overflow]
   48 |                 if (FLAGS_SET(flags, 1 << 40) && map[i])
      |                                        ^~
```

yuwata replied:
Thanks.

yuwata commented on src/network/networkd-address.c:
```diff
@@ -26,33 +26,30 @@
+                if (FLAGS_SET(flags, 1 << i) && map[i])
+                        if (!strextend_with_separator(
+                                            &str, ",",
+                                            family == AF_INET6 && (1 << i) == IFA_F_SECONDARY ? "temporary" : map[i]))
```
`UINT32_C(1)`, or `i == LOG2U(IFA_F_SECONDARY)`

keszybz replied:
Hmm, why would UINT32_C be better?

yuwata commented on src/network/networkd-link.c:
```diff
@@ -2660,50 +2660,47 @@ int link_call_getlink(Link *link, link_netlink_message_handler_t callback) {
-                if (flags & map[i].flag &&
-                    !strextend_with_separator(&str, ",", map[i].name))
-                        return -ENOMEM;
+                if (FLAGS_SET(flags, 1 << i) && map[i])
```
`UINT32_C(1)`

yuwata commented on src/network/networkd-route-util.c:
```diff
@@ -221,25 +221,22 @@ DEFINE_STRING_TABLE_LOOKUP_WITH_FALLBACK(route_protocol_full, int, UINT8_MAX);
+                [LOG2U(RTNH_F_OFFLOAD)] =    "offload"   , /* Nexthop is offloaded */
+                [LOG2U(RTNH_F_LINKDOWN)] =   "linkdown"  , /* carrier-down on nexthop */
+                [LOG2U(RTNH_F_UNRESOLVED)] = "unresolved", /* The entry is unresolved (ipmr) */
+                [LOG2U(RTNH_F_TRAP)] =       "trap"      , /* Nexthop is trapping packets */
```
Also here. Alignment is spurious.

yuwata commented on src/network/networkd-route-util.c:
```diff
@@ -221,25 +221,22 @@ DEFINE_STRING_TABLE_LOOKUP_WITH_FALLBACK(route_protocol_full, int, UINT8_MAX);
-                if (flags & map[i].flag &&
-                    !strextend_with_separator(&str, ",", map[i].name))
-                        return -ENOMEM;
+                if (FLAGS_SET(flags, 1 << i) && map[i])
```
`UINT32_C(1)`

yuwata commented on src/network/networkd-util.c:
```diff
@@ -24,25 +24,25 @@ static const char * const network_config_source_table[_NETWORK_CONFIG_SOURCE_MAX
-                    !strextend_with_separator(&buf, ",", map[i].str))
-                        return -ENOMEM;
+        for (size_t i = 0; i < ELEMENTSOF(states); i++)
+                if (FLAGS_SET(s, 1 << i)) {
```
Also here.

yuwata commented on src/network/test-networkd-util.c:
```diff
@@ -0,0 +1,19 @@
+#include "tests.h"
+
+TEST(network_config_state_to_string_alloc) {
+        for (unsigned i = 1; i <= NETWORK_CONFIG_STATE_REMOVING; i = (i << 1)) {
```
`i <<= 1`?

yuwata reviewed (approved):
Looks nice! Several minor comments.

keszybz commented:
Rebased and updated. I didn't change the double eval (see comment above), and UINT32_C, because I don't see the benefit (maybe I'm missing something).

poettering commented on src/basic/util.h:
```diff
@@ -22,32 +22,36 @@ void in_initrd_force(bool value);
+/* Note: log2(0) == log2(1) == 0 here and below. */
+
+/* Note: double evaluation of x */
+#define LOG2ULL(x) (x > 1 ? (unsigned) __builtin_clzll(x) ^ 63U : 0)
```
please enclose x in ()

poettering commented on src/basic/util.h:
```diff
@@ -22,32 +22,36 @@ void in_initrd_force(bool value);
-static inline unsigned log2i(int x) {
-        assert(x > 0);
+/* Note: double evaluation of x */
+#define LOG2U(x) (x > 1 ? __SIZEOF_INT__ * 8 - __builtin_clz(x) - 1 : 0)
```
ditto

keszybz commented:
OK, I updated with `__builtin_choose_expr(__builtin_constant_p(x), CONST_LOG2ULL(x), NONCONST_LOG2ULL(x))` as suggested by @medhefgo, and added () as requested by @poettering.

keszybz commented:
focal-s390x: TEST-37-RUNTIMEDIRECTORYPRESERVE:   FAIL     (1201 s)
